Major keyspace compaction task

Deexie opened the pull request:
Task manager task implementation that covers the major keyspace
compaction which can be start through /storage_service/keyspace_compaction/
api.

Deexie commented:
Draft until https://github.com/scylladb/scylladb/pull/12635 gets merged.

Deexie commented:
@bhalevy review please

bhalevy commented on api/storage_service.hh:
```diff
@@ -13,6 +13,7 @@
 #include <seastar/core/sharded.hh>
 #include "api.hh"
 #include "db/data_listeners.hh"
+#include "../table_info.hh"
```
That better be "api/table_info.hh" so it wouldn't be relative to this source location

bhalevy replied:
And there's already a definition of it in api/storage_service.hh that can be included here....

bhalevy commented on api/storage_service.cc:
```diff
@@ -668,21 +668,14 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
+
+        auto& cm = db.local().get_compaction_manager();
+        auto compaction_module_ptr = cm.get_compaction_module().shared_from_this();
+        auto task_impl_ptr = std::make_unique<major_keyspace_compaction_task_impl>(compaction_module_ptr, keyspace, db, table_infos);
```
keyspace and table_infos can be moved i to the task impl, no?

Deexie replied:
Not really due to that:
https://github.com/scylladb/scylladb/blob/6a2ef15365a2a066c02f38f7913d739657ba21fc/api/storage_service.cc#L680
Keyspace can be moved and then got from task, but table_infos should stay as is I think.

bhalevy commented on api/storage_service.cc:
```diff
@@ -668,21 +668,14 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
-                    return t.compact_all_sstables();
-                });
-            });
+            co_return task->done();
```
should probably be `co_return co_await task->done()`

Deexie replied:
It won't implicitly cast to `json_void`

Deexie commented:
- moved table_info.hh from root to api/
- moved keyspace in `ss::force_keyspace_compaction`

Deexie commented:
- moved table_info.hh from root to api/
- moved keyspace in ss::force_keyspace_compaction

bhalevy commented on tasks/task_manager.cc:
```diff
@@ -278,10 +278,19 @@ future<task_manager::task_ptr> task_manager::module::make_task(task::task_impl_p
+task_manager::task_manager() noexcept
+    : _update_task_ttl_action([this] { return update_task_ttl(); })
+    , _task_ttl_observer(_cfg.task_ttl.observe(_update_task_ttl_action.make_observer()))
+    , _task_ttl(_cfg.task_ttl.get())
```
why bother? `_cfg` is initialized to the default that has no value, no?

Deexie commented:
- rebased on master
- get_compaction_module -> get_task_manager_module
- moved major_keyspace_compaction_task_impl::run definition to compaction/task_manager_module.cc
	- copied run_on_existing_tables
	- moved operator<< for table_info

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4259/

bhalevy commented on compaction/task_manager_module.cc:
```diff
@@ -7,6 +7,33 @@
+    return os << "table{name=" << ti.name << ", id=" << ti.id << "}";
+}
+
+} // namespace std
```
I see no reason this would be implemented under `compaction/`.
there's already an implementation in api/storage_service.cc, no?
You can just declare it in the header file.

bhalevy commented on compaction/task_manager_module.cc:
```diff
@@ -7,6 +7,33 @@
+            co_await coroutine::return_exception_ptr(std::move(ex));
+        }
+    }
+}
```
this is a generic function that can be defined in api/storage_service.cc
or, we can create api/utils.{cc,hh} for generic useful stuff

Deexie replied:
Can't we move that out of api/? Eventually it will be used only in compaction/task_manager_module.cc.

If it stays in api/ then:
- we cannot keep definition in source file as api is not compiled for tests
- keeping definition in api/utils.hh (or any header) will make us include `replica/database.hh` which is heavy and should probably be avoided

I don't really know where is the right place for that.

bhalevy replied:
utils/api.{cc,hh} can host the generic stuff used by the api layer so it can be shared with other parts of scylla.
but I don't want such generic stuff to be buried here since it may very well be reused in other rest api calls.

raphaelsc replied:
looks like it was later agreed that it could live in compaction/. never mind then.

Deexie commented:
- added utils/api.{cc,hh}
- added major_compaction_task_impl class

bhalevy commented:
looks much better.
please just add a basic test

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4282/

Deexie commented:
- test added

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4317/

bhalevy commented on test/rest_api/test_compaction_task.py:
```diff
@@ -0,0 +1,36 @@
+from util import new_test_table, new_test_keyspace
+from rest_util import set_tmp_task_ttl
+
+def wait_for_task(rest_api, task_id):
```
since this helper isn't specific to compaction let's find some common place for it so it could be reused in many task_manager related tests.

Deexie replied:
I opened a PR with helpers to all task manager api calls kept together: https://github.com/scylladb/scylladb/pull/12844

Deexie commented:
- rebased on master
	- use task_manager_utils.py in test
	- allow test_compaction_task.py in release mode

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4512/

bhalevy commented on compaction/task_manager_module.hh:
```diff
@@ -8,10 +8,72 @@
+    }
+
+    virtual std::string type() const override {
+        // FIXME: implement
```
why is it not implemented yet?

Deexie commented:
- change compaction_task_impl::type() to pure virtual method

bhalevy reviewed (approved):
@raphaelsc please review too

denesb commented on utils/api.hh:
```diff
@@ -0,0 +1,31 @@
+/*
```
I don't understand why this file is called "api.hh"? We should name files based on their content, not where that content happened to live before.

denesb replied:
`utils/` in general is a place for low-level utility stuff that doesn't depend on any other parts of scylla. The content of this file however is higher level, as it works with database and tables.

Deexie replied:
How about `service/utils.hh`?

denesb replied:
I don't think this code is worth trying to make reusable, it is so simple, you can just duplicate it.  Code reuse is most of the time great, but not every single time. Here it would lead to compaction code including stuff from api or service namespace, which is not desired.

denesb commented on compaction/task_manager_module.hh:
```diff
@@ -31,6 +31,27 @@ protected:
+
+    virtual std::string type() const override {
+        return "major compaction";
+    }
```
How is this going to be used? Would it make sense to have an enum here instead of a string?

Deexie replied:
It is used in task manager api, from which user gets stringified status (containing type). We could have an enum type attribute, which is serialized in type() method. I am not sure it's worth it.

denesb replied:
Ok, it might not worth converting to an enum then.

denesb commented on compaction/task_manager_module.hh:
```diff
@@ -52,6 +54,23 @@ protected:
+class major_keyspace_compaction_task_impl : public major_compaction_task_impl {
+private:
+    sharded<replica::database>& _db;
+    std::vector<api::table_info> _table_infos;
```
Also here, please use standard type (table id), isntead of some api:: type.

Deexie commented:
- deleted utils/api.{hh, cc}
- copied run_on_existing_tables to compaction/task_manager_module.cc
- used table_id instead of api::table_info

denesb reviewed (approved):
LGTM

@bhalevy I asked @Deexie to remove utils/api.{hh,cc} and instead replicate the `run_on_existing_tables()` method into` compaction/, so we can avoid introducing dependency of compaction on api stuff. It is not worth it just to share such a simple function.

bhalevy commented:
> LGTM
> 
> @bhalevy I asked @Deexie to remove utils/api.{hh,cc} and instead replicate the `run_on_existing_tables()` method into compaction/, so we can avoid introducing dependency of compaction on api stuff. It is not worth it just to share such a simple function.

OK, I trust your judgement.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4579/
